README.md

```diff
@@ -1,18 +1,129 @@
+The exception effect adds the possibility to exit a computation preemptively with an exception. Note that the exceptions from this library are handled by the programmer and have nothing to do with exceptions thrown inside the Haskell run-time.
+
+```haskell
+throwError :: Member (Exc e) r => e -> Eff r a
+runError :: Eff (Exc e ': r) a -> Eff r (Either e a)
+```
+
+An exception can be thrown using the `throwError` function. Its return type is `Eff r a` with an arbitrary type `a`. When handling the effect, the result-type changes to `Either e a` instead of
+just `a`. This indicates how the effect is handled: The returned value is either the thrown exception or the value returned.
+
+### The State Effect
+
+The state effect provides readable and writable state during a computation.
+
+```haskell
+get :: Member (State s) r => Eff r s
+put :: Member (State s) r => s -> Eff r ()
+modify :: Member (State s) r => (s -> s) -> Eff r ()
+runState :: s -> Eff (State s ': r) a -> Eff r (a, s)
+```
 
-  * Effects can be added, removed, and interwoven without changes to code not
-    dealing with those effects.
+There are different functions for programming with the state-effect. The `get` functions accesses the current state and makes it usable within the further computation. The `put` function sets the state to the given value. `modify` updates the state using a mapping function.
+
+The state-effect is handled using `runState` function. It takes the initial state as argument and returns the final state and effect-result.
+
+### The Reader Effect
+
+The reader effect provides an environment that can be read. Sometimes it is considered as read-only state that stays same during the computation.
+
+```haskell
+ask :: Member (Reader e) r => e -> Eff r e
+runReader :: e -> Eff (Reader e ': r) a -> Eff r a
+```
+
+The environment given to the handle the reader effect is the one given during the computation if asked for.
+
+There are two variants of readers: strict and lazy. Each has its own module and provide the same interface. By importing one or the other, it can be controlled if the reader is strict or lazy in its environment argument.
+
+### The Writer Effect
+
+The writer effect allows to output messages during a computation. It is sometimes referred to as write-only state, which gets retrieved.
+
+```haskell
+tell :: Member (Writer w) r => w -> Eff r ()
+runWriter :: (w -> b -> b) -> b -> Eff (Writer w ': r) a -> Eff r (a, b)
+runListWriter :: Eff (Writer w ': r) a -> Eff r (a, [w])
+```
+
+running a writer can be done in several ways. The most general function is `runWriter` that folds over all written values. However, if you only want to collect the the values written, the `runListWriter` function does that.
+
+Note that compared to mtl, the value written has no Monoid constraint on it and can be collected in any way.
+
+### Using multiple effects
+
+The main benefit of this library is that multiple effects can be included without much changes necessary.
```

Comment:
@power-fungus this section could benefit from a concrete example. i.e., haskell code to accompany the description